### PR TITLE
fix out of range access, minor performance improvement

### PR DESCRIPTION
**Pull request by rabauke**
The current code leads to errors in gthumb in Ubuntu 20.4 and other applications:
```
terminate called after throwing an instance of 'std::out_of_range'
  what(): basic_string::at: __n (which is 19) >= this->size() (which is 19)
```
See also https://bugs.launchpad.net/ubuntu/+source/exiv2/+bug/1942799

### CODE REVIEW

File: src/tags_int.cpp [hunk 1]
```diff
@@ -3265,8 +3265,8 @@ namespace Exiv2 {
         }
```
Comment (rabauke) [inline src/tags_int.cpp:3266]: `value` has 19 or 20 characters.

File: src/tags_int.cpp [hunk 2]
```diff
@@ -3265,8 +3265,8 @@ namespace Exiv2 {
...
 
         std::string stringValue = value.toString();
-        if (stringValue.at(19) == 'Z') {
```
Comment (rabauke) [inline src/tags_int.cpp:3268]: Throws an "out of bound" exception if `stringValue` has only 19 characters.

File: src/tags_int.cpp [hunk 3]
```diff
@@ -3265,8 +3265,8 @@ namespace Exiv2 {
...
-            stringValue = stringValue.substr(0, 19);
```
Comment (rabauke) [inline src/tags_int.cpp:3269]: Unneeded creation of a _new_ string, which involves copying and memory allocation.

**Comment by kevinbackhouse** [comment]
Thanks for this @rabauke! Do you have an example image that triggers this, so that we can add a test?

**Comment by kevinbackhouse** [comment]
@clanmills: do you think we should include this fix in 0.27.5?

**Comment by rabauke** [comment]
@kevinbackhouse Here is a suitable image with embedded xmp data and time field for testing:
A1_05759

**Comment by clanmills** [comment]
@kevinbackhouse If you believe this fix merits having RC4 and delaying 0.27.5 until mid-October, then that's we should do.  

Thanks to a discussion in #1915, we could remove the fix for #1857 that's in RC3.  I think you have other recent security fixes which could be rolled into RC4.

Without studying the severity of this vulnerability, I don't know if we should delay or not.  I don't think we want to get into an endless loop of release candidates.

**Comment by kevinbackhouse** [comment]
@clanmills: I think this is a regular bug, rather than a security vulnerability, but it does sound like it was discovered on a genuine use-case, so that's why it might be worth back-porting. The fix looks low-risk to me.

If we're doing an RC4 then we could consider including #1902 and #1910. Those are long-running loops that could maybe be used to cause a denial of service. I don't have anything else that's worth back-porting.

**Comment by kevinbackhouse** [comment]
This is reproducible like this:

```bash
exiv2 -px ~/Downloads/134555125-dc532f12-4ca6-4829-84dd-8f8e63f56c3c.jpg
```

Error message:

```
Xmp.exif.DateTimeOriginal                    XmpText    19  Uncaught exception: basic_string::at: __n (which is 19) >= this->size() (which is 19)
```

Prior to #1786 this was an out-of-bounds read, which was probably invisible to the user. Now it throws exception, which means that it's no longer a vulnerability, but it's become a visible bug.

**Comment by kevinbackhouse** [comment]
I created a PR to add a test: https://github.com/rabauke/exiv2/pull/1

**Comment by rabauke** [comment]
@kevinbackhouse wrote

> Prior to #1786 this was an out-of-bounds read, which was probably invisible to the user. Now it throws exception, which means that it's no longer a vulnerability, but it's become a visible bug.

I tend to agree here. Some security researchers, however, might argue that the uncaught exception (the bug) may allow denial of service attacks by sending files with xmp data to an application that uses the exiv2 library. Opening such a file can lead to application crashes. 

Currently, I am just annoyed that gthumb has become unusable for for me due to this bug (until I fixed it locally).

**Comment by clanmills** [comment]
Gentlemen:

Thanks for working on this.  I'm on vacation at the moment.  v0.27.5 GM was scheduled for 2021-09-30 (next Thursday).  Let's go for RC4 on 2021-09-30 and move v0.27.5 GM to 2021-10-22.

**Comment by kevinbackhouse** [comment]
@clanmills: Thanks, that sounds like a good plan. I'll use mergify to back-port the pull requests. By the way, you mean RC3, don't you. (RC2 is the most recent release that I know of.)

**Review by kevinbackhouse — APPROVED** [review]

**Comment by kevinbackhouse** [comment]
@Mergifyio backport 0.27-maintenance

**Comment by clanmills** [comment]
Right.  I made an RC of my RCs.  You are right.  RC3.

I thought today about _taking a chance_ and  pushing the changes in waiting and declaring GM.  RC4 is a negative message.  However, I'm 100% fine with RC3 on 2021-09-30 and GM on 2021-10-22.  

I'm going to yank #1857 and redo it with the great ideas from @evanmiller in  #1915.  Team work is really good stuff.  When I do that, you'll probably want to merge my PR into main.

I'm on vacation at the moment, and will deal with RC3 next week.

**Pull request merged**